src/Command/PurgeCommand.php

```diff
@@ -0,0 +1,249 @@
+
+                // Add custom logic
+                if (isset($GLOBALS['TL_HOOKS']['postLeadsPurge']) && is_array($GLOBALS['TL_HOOKS']['postLeadsPurge'])) {
+                    foreach ($GLOBALS['TL_HOOKS']['postLeadsPurge'] as $callback) {
+                        if (is_array($callback)) {
+                            System::importStatic($callback[0])->{$callback[1]}(
+                                $masterForm,
+                                $leads,
+                                $leadsData,
+                                $deletedUploads
+                            );
+                        } elseif (is_callable($callback)) {
+                            $callback($masterForm, $leads, $leadsData, $deletedUploads);
+                        }
+                    }
+                }
+
+                $purged = true;
+            }
+        }
+
+        return $purged;
+    }
+
+    /**
+     * @param $masterForm
+     * @return array
+     */
+    private function getAllLeads($masterForm)
+    {
+        $leads = [];
+
+        $rows = $this->db->fetchAll(
+            "SELECT * FROM tl_lead WHERE master_id=? AND created<?",
+            [$masterForm['id'], (time() - (int)$masterForm['leadPeriod'])]
+        );
+
+        foreach ($rows as $row) {
+            $leads[$row['id']] = $row;
+        }
+
+        return $leads;
+    }
+
+    /**
+     * @param $masterForm
+     * @param $leadsIds
+     * @return array
+     */
+    private function getAllLeadsData($masterForm, $leadsIds)
+    {
+        $leadsData = [];
+
+        if (!empty($leadsIds)) {
+            $rows = $this->db->fetchAll(
+                "SELECT d.*, f.type AS field_type FROM tl_lead_data d
+                      LEFT JOIN tl_form_field f ON d.field_id = f.id
+                      WHERE d.pid IN(".$leadsIds.") AND d.tstamp<?",
+                [(time() - (int)$masterForm['leadPeriod'])]
+            );
+
+            foreach ($rows as $row) {
+                $leadsData[$row['id']] = $row;
+            }
+        }
+
+        return $leadsData;
+    }
+
+
+    /**
+     * @param $leadsData
+     * @return array
+     */
+    private function purgeUploads($leadsData)
+    {
+        $files = [];
+        if (!empty($leadsData)) {
+            foreach ($leadsData as $data) {
+                if ('upload' === $data['field_type']) {
+                    $files[$data['id']] = $this->purgeUpload($data['value']);
+                }
+            }
+        }
+
+        return $files;
+    }
+
+    /**
+     * @param $value
+     * @return FilesModel|null
+     */
+    private function purgeUpload($value)
+    {
+        if (!Validator::isUuid($value)) {
+            return null;
+        }
+
+        $filesModel = FilesModel::findByUUid($value);
+
+        if (null !== $filesModel) {
+            try {
+                $file = new File($filesModel->path);
```

Comment:
We should try to use as little old code as possible. PHP (`SplFileObject`) and Symfony (`Filesystem`) already provide excellent tools to work with files, there's no need to use old code here. Bot, PHP and Symfony are well tested. Contao's old libraries are not and should thus be avoided wherever you can (you cannot everywhere yet but use the better tools for the job if you can 😄 ).